File: apps/macos/Sources/TrimletCore/KeyframeIndex.swift
import Foundation

public struct FastCutCandidate: Equatable, Sendable {
    public let start: Double
    public let end: Double
    public let requestedStart: Double
    public let requestedEnd: Double

    public var startDelta: Double { start - requestedStart }
    public var endDelta: Double { end - requestedEnd }
}

public struct KeyframeIndex: Equatable, Sendable {
    public let duration: Double
    public let keyframes: [Double]

    public init(duration: Double, keyframes: [Double]) {
        self.duration = max(0, duration)
        self.keyframes = Array(Set(keyframes.filter { $0.isFinite && $0 >= 0 && $0 <= duration + 0.001 }))
            .sorted()
    }

    public func fastCandidate(for range: TrimRange) -> FastCutCandidate? {
        guard let requestedStart = range.inPoint,
              let requestedEnd = range.outPoint,
              requestedEnd > requestedStart,
              let start = keyframes.last(where: { $0 <= requestedStart + 0.000_001 }) else {
            return nil
        }

        let end = keyframes.first(where: { $0 >= requestedEnd - 0.000_001 }) ?? duration
        guard end > start else { return nil }
        return FastCutCandidate(
            start: start,
            end: min(end, duration),
            requestedStart: requestedStart,
            requestedEnd: requestedEnd
        )
    }
}

File: apps/windows/src/Trimlet.Media/KeyframeIndex.cs
namespace Trimlet.Media;

public sealed record FastCutCandidate(
    MediaTimestamp Start,
    MediaTimestamp End,
    MediaTimestamp RequestedStart,
    MediaTimestamp RequestedEnd)
{
    public double StartDeltaSeconds => Start.TotalSeconds - RequestedStart.TotalSeconds;
    public double EndDeltaSeconds => End.TotalSeconds - RequestedEnd.TotalSeconds;
}

public sealed class KeyframeIndex
{
    public KeyframeIndex(MediaTimestamp duration, IEnumerable<MediaTimestamp> keyframes)
    {
        Duration = duration;
        Keyframes = keyframes
            .Where(value => value <= duration)
            .Distinct()
            .OrderBy(value => value)
            .ToArray();
    }

    public MediaTimestamp Duration { get; }
    public IReadOnlyList<MediaTimestamp> Keyframes { get; }

    public FastCutCandidate? FastCandidate(TrimRange range)
    {
        var start = Keyframes.LastOrDefault(value => value <= range.In);
        if (start.Timescale <= 0)
        {
            return null;
        }

        var end = Keyframes.FirstOrDefault(value => value >= range.Out);
        if (end.Timescale <= 0)
        {
            end = Duration;
        }

        return end > start
            ? new FastCutCandidate(start, end, range.In, range.Out)
            : null;
    }
}
